src/app/meshes/data/index.ts

```diff
@@ -0,0 +1,84 @@
+import { getDataplaneStatusCounts } from '@/app/data-planes/data'
+import { getServiceTypeCount } from '@/app/services/data'
+import type { PaginatedApiListResponse } from '@/types/api.d'
+import type {
+  Backend,
+  DataPlaneProxyStatus as DataplaneStatusCount,
+  MeshBackend,
+  Mesh as PartialMesh,
+  MeshInsight as PartialMeshInsight,
+  ServiceStatus as ServiceTypeCount,
+} from '@/types/index.d'
+
+export type Mesh = PartialMesh & {
+  mtlsBackend: Backend | undefined
+  metricsBackend: Backend | undefined
+}
+
+export type MeshInsight = PartialMeshInsight & {
+  dataplanes: Required<DataplaneStatusCount>
+  dataplanesByType: {
+    standard: Required<DataplaneStatusCount>
+    gateway: Required<DataplaneStatusCount>
+    gatewayBuiltin: Required<DataplaneStatusCount>
+    gatewayDelegated: Required<DataplaneStatusCount>
+  }
+  services: Required<ServiceTypeCount>
+  totalPolicyCount: number
+}
+
+export const Mesh = {
+  fromObject(partialMesh: PartialMesh): Mesh {
+    const mtlsBackend = getBackend(partialMesh.mtls)
+    const metricsBackend = getBackend(partialMesh.metrics)
+
+    return {
+      ...partialMesh,
+      mtlsBackend,
+      metricsBackend,
+    }
+  },
+}
+
+export const MeshInsight = {
+  fromObject(partialMeshInsight: PartialMeshInsight): MeshInsight {
+    const dataplanes = getDataplaneStatusCounts(partialMeshInsight.dataplanes)
+    const dataplanesByType = {
+      standard: getDataplaneStatusCounts(partialMeshInsight.dataplanesByType.standard),
+      gateway: getDataplaneStatusCounts(partialMeshInsight.dataplanesByType.gateway),
+      gatewayBuiltin: getDataplaneStatusCounts(partialMeshInsight.dataplanesByType.gatewayBuiltin),
+      gatewayDelegated: getDataplaneStatusCounts(partialMeshInsight.dataplanesByType.gatewayDelegated),
+    }
+    const services = getServiceTypeCount(partialMeshInsight.services)
+    const totalPolicyCount = Object.values(partialMeshInsight.policies ?? {}).reduce((total, stat) => total + stat.total, 0)
+
+    return {
+      ...partialMeshInsight,
+      dataplanes,
+      dataplanesByType,
+      services,
+      totalPolicyCount,
+    }
+  },
+
+  fromCollection(partialMeshInsights: PaginatedApiListResponse<PartialMeshInsight>): PaginatedApiListResponse<MeshInsight> {
+    return {
+      ...partialMeshInsights,
+      items: Array.isArray(partialMeshInsights.items)
+        ? partialMeshInsights.items.map((partialMeshInsight) => MeshInsight.fromObject(partialMeshInsight))
+        : [],
+    }
+  },
+}
+
+function getBackend(meshBackend?: MeshBackend): Backend | undefined {
+  if (meshBackend?.enabledBackend && Array.isArray(meshBackend.backends)) {
+    const enabledBackend = meshBackend.backends.find((backend) => backend.name === meshBackend.enabledBackend)
+
+    if (enabledBackend !== undefined) {
+      return enabledBackend
+    }
```

Comment:
> Unless I missed something, seeing as `find` returns the thing or `undefined` you should be able to just return the result of `find` here.

Correct, I changed this.

> I think we should make `meshBackend.backends` always an array before we get here also. Once those two are done, seeing as you are pretty much just calling `find` the util function is pretty redundant.

I’m gonna leave this as-is for now. Changing this so that mesh backends that are set  always have `backends` just in order to simplify the `getBackend` function would ultimately be more work/code right now and we don’t access this anywhere else.

Reply:
Yep no prob, thanks for the simplification, resolving